clang/lib/CIR/Dialect/Transforms/HoistAllocas.cpp

```diff
@@ -48,8 +48,14 @@ static void process(cir::FuncOp func) {
 
   mlir::Operation *insertPoint = &*entryBlock.begin();
 
-  for (auto alloca : allocas)
+  for (auto alloca : allocas) {
     alloca->moveBefore(insertPoint);
+    if (alloca.getConstant()) {
+      // For now, we remove the const flag on nested allocas for constant
+      // variables when hoisting them.
```

Comment:
In theory the alloca is still const, this is a workaround for the invariant side of things that are going to be generated later, in theory we could be propagating another attribute that has the specific semantic and delete it instead. I'm not sure it's worth adding that right now though. Can you explain that in the comment instead?

Reply:
> in theory we could be propagating another attribute that has the specific semantic and delete it instead

I'm wondering recently if we could leverage the still-missing CIR lifetime markers as pointed out in #1129 . The rough idea is the constness property would only hold in the program points between the lifetime start and end markers, and then we don't have to remove the `const` flag here. What do you think?

Anyway, this would not block this PR. Will update the comment here later.

Reply:
Looks like as is the effectiveness of this PR is reduced because basically anything that's not at the function level scope is going to be hoisted, right?

> What do you think?

I'm a bit on the fence, that would require LowerToLLVM to have to check for dominance in order to properly emit it. Have you gave some thought on the implementation details?